overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -69,20 +69,31 @@ func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([
 
 var servicestateControl = servicestate.Control
 
-func queueCommand(context *hookstate.Context, ts *state.TaskSet) {
-	// queue command task after all existing tasks of the hook's change
+func queueCommand(context *hookstate.Context, ts *state.TaskSet) error {
+	hookTask, ok := context.Task()
+	if !ok {
+		return fmt.Errorf("attempted to queue command with ephemeral context")
+	}
+
 	st := context.State()
 	st.Lock()
 	defer st.Unlock()
 
-	task, ok := context.Task()
-	if !ok {
-		panic("attempted to queue command with ephemeral context")
+	change := hookTask.Change()
+	hookTaskLanes := hookTask.Lanes()
+	tasks := change.LaneTasks(hookTaskLanes...)
+	if len(hookTaskLanes) == 1 && hookTaskLanes[0] != 0 {
+		ts.JoinLane(hookTaskLanes[0])
+	} else {
+		for _, l := range hookTaskLanes {
+			ts.JoinLane(l)
+		}
 	}
-	change := task.Change()
-	tasks := change.Tasks()
 	ts.WaitAll(state.NewTaskSet(tasks...))
 	change.AddAll(ts)
+	st.EnsureBefore(0)
```

Comment:
ah, good catch! this probably merits a comment, maybe something like:

// as this can be run from what was originally the last task of a change, make sure the tasks added to the change are considered immediately

Reply:
Comment added.